apollo-router/src/plugins/router_limits/mod.rs

```diff
@@ -0,0 +1,161 @@
+
+use std::num::NonZeroU64;
+use std::time::Duration;
+
+use http::StatusCode;
+use schemars::JsonSchema;
+use serde::Deserialize;
+use serde::Serialize;
+use tower::limit::RateLimitLayer;
+use tower::load_shed::error::Overloaded;
+use tower::BoxError;
+use tower::ServiceBuilder;
+use tower::ServiceExt;
+
+use crate::graphql;
+use crate::layers::ServiceBuilderExt;
+use crate::plugin::PluginInit;
+use crate::plugin::PluginPrivate;
+use crate::services::router;
+use crate::services::RouterResponse;
+
+#[derive(PartialEq, Debug, Clone, Deserialize, JsonSchema)]
+#[serde(deny_unknown_fields)]
+/// The limits placed on a router in virtue of what's in a user's license
+pub(crate) struct RouterLimits {
+    /// Transactions per second allowed based on license tier
+    pub(crate) tps: Option<TpsLimitConf>,
+}
+
+#[derive(PartialEq, Debug, Clone, Deserialize, JsonSchema)]
+#[serde(deny_unknown_fields)]
+/// Configuration for transactions per second
+pub(crate) struct TpsLimitConf {
+    /// The number of operations allowed during a certain interval
+    pub(crate) capacity: NonZeroU64,
+    /// The interval as specied in the user's license; this is in milliseconds
+    #[serde(deserialize_with = "humantime_serde::deserialize")]
+    #[schemars(with = "String")]
+    pub(crate) interval: Duration,
+}
+
+#[derive(Debug, Default, Deserialize, JsonSchema, Serialize)]
+pub(crate) struct RouterLimitsConfig {}
+
+#[async_trait::async_trait]
+impl PluginPrivate for RouterLimits {
+    type Config = RouterLimitsConfig;
+
+    async fn new(init: PluginInit<Self::Config>) -> Result<Self, BoxError> {
+        let tps = init.license.get_limits().and_then(|limits| {
+            limits.tps.and_then(|tps| {
+                NonZeroU64::new(tps.capacity as u64).map(|capacity| TpsLimitConf {
+                    capacity,
+                    interval: tps.interval,
+                })
+            })
+        });
+
+        Ok(Self { tps })
+    }
+
+    fn router_service(&self, service: router::BoxService) -> router::BoxService {
+        ServiceBuilder::new()
+            .map_future_with_request_data(
+                |req: &router::Request| req.context.clone(),
+                move |ctx, future| async {
+                    let response: Result<RouterResponse, BoxError> = future.await;
+                    match response {
+                        Ok(ok) => Ok(ok),
+                        Err(err) if err.is::<Overloaded>() => {
+                            let error = graphql::Error::builder()
+                                .message("Your request has been rate limited. You've reached the limits for the Free plan. Consider upgrading to a higher plan for increased limits.")
+                                .extension_code("ROUTER_FREE_PLAN_RATE_LIMIT_REACHED")
+                                .build();
+                            Ok(RouterResponse::error_builder()
```

Comment:
Either the router response needs to increment the error metrics or we need to increment the metrics here